### PR TITLE
RHOAIENG-209-Deprecate the use of Tekton ClusterTasks in any Pipelines

### PR DESCRIPTION
**Pull request by biswassri**
JIRA: [RHOAIENG-209](https://issues.redhat.com/browse/RHOAIENG-209)
## Description

Deprecating clustertasks and replacing with git-resolvers. The idea is we can currently pulling in pre-defined task yamls from the tekton catalog git repo. Ref: [Migration from clustertasks to tekton-resolvers ](https://www.redhat.com/en/blog/migration-from-clustertasks-to-tekton-resolvers-in-openshift-pipelines)

## How Has This Been Tested?

```
oc apply -k manifests/
oc create -f examples/tekton/aiedge-e2e/example-pipelineruns/git-fetch.tensorflow-housing.pipelinerun.yaml
```
## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by biswassri** [comment]
The change for the `build-container-image` is still something I'm figuring out. Using the same method as others, I'm getting errors. So still working on it.

**Comment by MarianMacik** [comment]
Rebase is needed

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:388]: I noticed that now it uses images from `quay.io`. Before it used cluster tasks installed by OpenShift Pipelines, which means it used images from `registry.redhat.io`. This may be the reason why `skopeo-copy` task is failing with:
```
level=fatal msg="initializing source docker://image-registry.openshift-image-registry.svc:5000/pipeline-test-tensorflow-housing/tensorflow-housing@sha256:2b6b85d6c3998ae2efb73c931755710bb7ac000925bc6c15479b6341ac3d1bff: reading manifest sha256:2b6b85d6c3998ae2efb73c931755710bb7ac000925bc6c15479b6341ac3d1bff in image-registry.openshift-image-registry.svc:5000/pipeline-test-tensorflow-housing/tensorflow-housing: authentication required"
```
So maybe with Tekton task the authentication is not working correctly with OpenShift Pipelines operator and there may be some incompatibility. I have rerun the PR check to see if it is persistent. We may need to use the tasks directly from RH to ensure the compatibility.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:109]: S3 version of the pipeline and GitOps pipeline should also have these, right?

Comment (biswassri) [reply manifests/pipelines/git-fetch-pipeline.yaml:109]: Yeap, I have that ready. Just haven't pushed it because of the buildah one.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (biswassri) [inline manifests/pipelines/s3-fetch-pipeline.yaml:325]: @MarianMacik I'll update this once this PR is merged in https://github.com/openshift-pipelines/task-openshift/pull/47

**Comment by MarianMacik** [comment]
There is an error `'[User error] workspace binding "images-url" does not match any declared workspace'` on the `skopeo-copy` task. Which is weird, given the previous `retrieve-build-image-info` task with exactly the same workspace binding passed. Probably an issue on the cluster with Persistence volumes?

**Comment by MarianMacik** [comment]
@biswassri I understand now. The message `'[User error] workspace binding "images-url" does not match any declared workspace'` means that our binding named images-url doesn't match any workspace which is declared on that task, in this case `skopeo-copy`.
I see that we use the provided task by the Pipelines operator and based on this docs: https://docs.openshift.com/pipelines/1.15/create/remote-pipelines-tasks-resolvers.html#op-taskref-skopeo-copy_remote-pipelines-tasks-resolvers
There were changes to the original `ClusterTask`:

Changes from the skopeo-copy ClusterTask:

- All parameter names were changed to uppercase.
- The VERBOSE parameter was added.
- The workspace name was changed from images-url to images_url.
- The SOURCE_DIGEST and DESTINATION_DIGEST results were added.

So we need do change that. Ideally unify across whole pipeline, this means change also the definition of the `retrieve-build-image-info` task we ship.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:390]: Pipeline is failing because of this. We should use the `images_url` workspace as before. The README of that Tekton task is missing the information about workspaces but the `images_url` workspace is still there and can be used as before, see [here](https://github.com/openshift-pipelines/tektoncd-catalog/blob/5a1664f1adf1f6cc1323c6cc2a2bdf390658defb/tasks/task-skopeo-copy/0.4.1/task-skopeo-copy.yaml#L30-L33).

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:393]: Also this VERBOSE config is not needed I think.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (biswassri) [reply manifests/pipelines/git-fetch-pipeline.yaml:390]: @MarianMacik  I think the pipeline is failing because of this: `level=fatal msg="Invalid source name $(tasks.retrieve-build-image-info.source_image_url.path): Invalid image name \"$(tasks.retrieve-build-image-info.source_image_url.path)\", expected colon-separated transport:reference"` I updated the retrieve-build-image-info task and I think that change is causing the above error and the pipeline failure.

Comment (MarianMacik) [reply manifests/pipelines/git-fetch-pipeline.yaml:390]: @biswassri But that is the thing, we shoudn't change the way we specify images in the retrieve-build-image-info task. That should stay the same as it was before, that is using the url.txt file. Or is there any reason for this change?

Comment (biswassri) [reply manifests/pipelines/git-fetch-pipeline.yaml:390]: Yes, If you see the new skopeo-copy script, they are no longer user URL.txt file. The new task has added params for source and destination URLs. [here](https://github.com/openshift-pipelines/tektoncd-catalog/blob/5a1664f1adf1f6cc1323c6cc2a2bdf390658defb/tasks/task-skopeo-copy/0.4.1/task-skopeo-copy.yaml#L35-L43)

**Comment by biswassri** [comment]
@MarianMacik @jackdelahunt This is ready for review. Please /lgtm whenever ready

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [reply manifests/pipelines/git-fetch-pipeline.yaml]: @biswassri It still has it. [This is the latest version 0.4.1](https://github.com/openshift-pipelines/tektoncd-catalog/blob/p/tasks/task-skopeo-copy/0.4.1/task-skopeo-copy.yaml#L106). It also still has the [workspace definition and mention of copying multiple images](https://github.com/openshift-pipelines/tektoncd-catalog/blob/p/tasks/task-skopeo-copy/0.4.1/task-skopeo-copy.yaml#L29-L33).

When I decode the linked skopeo-copy.sh script, I get this:

```bash
#!/usr/bin/env bash

shopt -s inherit_errexit
set -eu -o pipefail

source "$(dirname "${BASH_SOURCE[0]}")/common.sh"
source "$(dirname "${BASH_SOURCE[0]}")/skopeo-common.sh"

# Ensure the /tekton/home/.docker directory exists
mkdir -p /workspace/home/.docker

phase "Copying '${PARAMS_SOURCE_IMAGE_URL}' into '${PARAMS_DESTINATION_IMAGE_URL}'"

set -x

if [ -n "${PARAMS_SOURCE_IMAGE_URL}" ] && [ -n "${PARAMS_DESTINATION_IMAGE_URL}" ]; then
    skopeo copy ${SKOPEO_DEBUG_FLAG} \
        --src-tls-verify="${PARAMS_SRC_TLS_VERIFY}" \
        --dest-tls-verify="${PARAMS_DEST_TLS_VERIFY}" \
        "${PARAMS_SOURCE_IMAGE_URL}" \
        "${PARAMS_DESTINATION_IMAGE_URL}"
else
    # Function to copy multiple images.
    copyimages() {
        filename="${WORKSPACES_IMAGES_URL_PATH}/url.txt"
        while IFS= read -r line || [ -n "$line" ]
        do
            cmd=""
            for url in $line
            do
                cmd="$cmd $url"
            done
            read -ra SOURCE <<<"${cmd}"
            skopeo copy "${SOURCE[@]}" --src-tls-verify="${PARAMS_SRC_TLS_VERIFY}" --dest-tls-verify="${PARAMS_DEST_TLS_VERIFY}"
            echo "$cmd"
        done < "$filename"
    }

    copyimages
fi
```
which clearly shows the possibility to use the `url.txt` file.

It is available since version 0.3.1.

Comment (biswassri) [reply manifests/pipelines/git-fetch-pipeline.yaml]: @MarianMacik okay sorry, I actually hadn't decoded the skopeo-copy script. I'll revert the retrieve info script changes. But in that case, what would we populate `SOURCE_IMAGE_URL` and `DESTINATION_IMAGE_URL` with? I thought that this was to be populated as a task result from retrieve info.

Comment (MarianMacik) [reply manifests/pipelines/git-fetch-pipeline.yaml]: @biswassri I see. In that case they should be empty. Otherwise the url.txt file won't be taken into consideration.

File: manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml:31]: I think in this task we just need to change the workspace name to `images_url` and that's all.

**Review by MarianMacik — APPROVED** [review]

File: manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml:31]: I think here we just need to change this workspace name to `images_url` and that's all.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (jackdelahunt) [inline manifests/pipelines/git-fetch-pipeline.yaml:313]: I was going to say this is wrong but I checked and it is actually called p...

**Review by jackdelahunt — CHANGES_REQUESTED** [review]
This panic is getting triggered in the e2e-tests because there is a new possible state for the task https://github.com/opendatahub-io/ai-edge/blob/main/test/e2e-tests/tests/pipelines_test.go#L296-L309

Adding a new case called "ResolvingTaskRef" that returns `false, nil` should fix the issue

**Comment by jackdelahunt** [comment]
I am also getting other errors in the e2e-tests but that will need further de-bugging

File: manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml:76]: This is no longer relevant as we use the url.txt file.

File: manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml:70]: This script should be left as before.

File: manifests/tasks/retrieve-build-image-info/README.md
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/README.md:26]: This can be removed.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/s3-fetch-pipeline.yaml:356]: After that, this can be left empty:
```suggestion
    - name: SOURCE_IMAGE_URL
      value: ""
    - name: DESTINATION_IMAGE_URL
      value: ""
```

**Comment by jackdelahunt** [comment]
`git-fetch` e2e test passes but the `s3-fetch` pipeline is failing on the `test-container-deploy` step. I will try to debug this and see if is your changes

**Comment by adelton** [comment]
As a baseline for trying this change, I did run
```
$ oc apply -k manifests/
$ oc create -f examples/tekton/aiedge-e2e/example-pipelineruns/git-fetch.tensorflow-housing.pipelinerun.yaml
```
on the `main` branch checkout and got `fetch-model-git` failing with
```
{"level":"error","ts":1727270196.2212334,"caller":"git/git.go:53","msg":"Error running git [init /workspace/output//model_dir-tensorflow-housing/]: exit status 128\nfatal: cannot mkdir /workspace/output//model_dir-tensorflow-housing/: Permission denied\n","stacktrace":"github.com/tektoncd-catalog/git-clone/git-init/git.run\n\t/go/src/github.com/tektoncd-catalog/git-clone/image/git-init/git/git.go:53\ngithub.com/tektoncd-catalog/git-clone/git-init/git.Fetch\n\t/go/src/github.com/tektoncd-catalog/git-clone/image/git-init/git/git.go:88\nmain.main\n\t/go/src/github.com/tektoncd-catalog/git-clone/image/git-init/main.go:52\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:271"}
{"level":"fatal","ts":1727270196.221359,"caller":"git-init/main.go:53","msg":"Error fetching git repository: exit status 128","stacktrace":"main.main\n\t/go/src/github.com/tektoncd-catalog/git-clone/image/git-init/main.go:53\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:271"}
```

With this PR, `fetch-model-git` passes, and so does the rest of the pipeline apart from the push to registry, which was due to me not having credentials set right.

So on the functionality front, this is a big improvement as it actually restores a working state.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/s3-fetch-pipeline.yaml:369]:
```suggestion
    - name: images_url
```
As per [documentation](https://github.com/openshift-pipelines/tektoncd-catalog/tree/p/tasks/task-skopeo-copy/0.4.2#workspace).
This now needs to be reverted back as we use the new task and not the ClusterTask.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:400]:
```suggestion
    - name: images_url
```

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:387]: Same here.
```suggestion
    - name: SRC_TLS_VERIFY
      value: "true"
    - name: DEST_TLS_VERIFY
      value: "true"
```

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/s3-fetch-pipeline.yaml:356]: Now we can leave this as it was before - just explicitly mentioning TLSverify as true. Just name of the [parameters](https://github.com/openshift-pipelines/tektoncd-catalog/tree/p/tasks/task-skopeo-copy/0.4.2#params) are `SRC_TLS_VERIFY` and `DEST_TLS_VERIFY`.
```suggestion
    - name: SRC_TLS_VERIFY
      value: "true"
    - name: DEST_TLS_VERIFY
      value: "true"
```

File: manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml
Comment (MarianMacik) [inline manifests/tasks/retrieve-build-image-info/retrieve-build-image-info.yaml:8]:
```suggestion
  - name: images_url
```

**Review by MarianMacik — CHANGES_REQUESTED** [review]
It is failing because the workspace in the new `skopeo-copy` task is called `images_url`.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (biswassri) [reply manifests/pipelines/s3-fetch-pipeline.yaml:369]: Ah okay, makes sense! I was looking into logs but couldn't find anything. I'll update these changes.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (biswassri) [reply manifests/pipelines/s3-fetch-pipeline.yaml:356]: @MarianMacik Sure. I removed it because these are by default true. But i can add them.

File: manifests/pipelines/s3-fetch-pipeline.yaml
Comment (MarianMacik) [reply manifests/pipelines/s3-fetch-pipeline.yaml:369]: @biswassri You can see the log [here](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/opendatahub-io_ai-edge/290/pull-ci-opendatahub-io-ai-edge-main-test-ai-edge/1841494657213665280/artifacts/test-ai-edge/test-ai-edge/artifacts/s3-fetch-bike-rentals-auto-ml-s78w6/). 
Search for `failed` in the taskRuns.txt.

**Comment by biswassri** [comment]
@MarianMacik Resolved all the above requested changes. Should be good to merge after openshift-ci issues resolve.

**Comment by adelton** [comment]
> 
> But I don't think it is caused by this PR as it does not touch it. Will investigate some more.

I've reproduced the issue on the main branch as well.

I had to do
```
$ oc apply -f - < examples/tekton/aiedge-e2e/test-data/tensorflow-housing-test-data-cm.yaml
```
to workaround the issue. It is weird because we don't have this documented in https://github.com/opendatahub-io/ai-edge/blob/main/manifests/README.md so in the past it was not needed as separate step.

**Comment by adelton** [comment]
The next issue I'm hitting is
```
pod "tensorflow-housing-1-6d6c48cd77-cwp82" deleted
service "tensorflow-housing-1" deleted
deployment.apps "tensorflow-housing-1" deleted
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
Error from server (Forbidden): applications.app.k8s.io is forbidden: User "system:serviceaccount:q-1:pipeline" cannot list resource "applications" in API group "app.k8s.io" in the namespace "q-1"
```
for the git-fetch-tensorflow-housing-wgdvk-stop-deployment Task. But as above, I also see this on the main branch for the stop-deployment ClusterTask, so not specific to this PR.

File: manifests/pipelines/git-fetch-pipeline.yaml
Comment (MarianMacik) [inline manifests/pipelines/git-fetch-pipeline.yaml:387]: I would remove these as they don't add any value and the values are default anyway.

**Review by MarianMacik — COMMENTED** [review]
Last thing @biswassri.
It is not needed to properly function but I think we can change it to as it was before to reduce the number of changes, wdyt?

**Comment by MarianMacik** [comment]
> The next issue I'm hitting is now
> 
> ```
> pod "tensorflow-housing-1-6d6c48cd77-cwp82" deleted
> service "tensorflow-housing-1" deleted
> deployment.apps "tensorflow-housing-1" deleted
> Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
> Error from server (Forbidden): applications.app.k8s.io is forbidden: User "system:serviceaccount:q-1:pipeline" cannot list resource "applications" in API group "app.k8s.io" in the namespace "q-1"
> ```
> 
> for the git-fetch-tensorflow-housing-wgdvk-stop-deployment Task when the `oc delete all --selector=app=tensorflow-housing-1` command is run. But as above, I also see this on the main branch for the stop-deployment ClusterTask, so not specific to this PR.

@adelton Yes, you are right. We have found this issue as well when using delete option of the `upon-end` parameter. We use [`keep`](https://github.com/opendatahub-io/ai-edge/blob/main/test/shell-pipeline-tests/openvino-tensorflow-housing/pipelines-test-openvino-tensorflow-housing.sh#L43) in PR check to obtain the logs of the deployment.

**Comment by MarianMacik** [comment]
> > But I don't think it is caused by this PR as it does not touch it. Will investigate some more.
> 
> I've reproduced the issue on the main branch as well.
> 
> I had to do
> 
> ```
> $ oc apply -f - < examples/tekton/aiedge-e2e/test-data/tensorflow-housing-test-data-cm.yaml
> ```
> 
> to workaround the issue. It is weird because we don't have this documented in https://github.com/opendatahub-io/ai-edge/blob/main/manifests/README.md so in the past it was not needed as separate step.

This is documented [here](https://github.com/opendatahub-io/ai-edge/tree/main/manifests#data-for-testing-the-model-inferencing-endpoint).

**Comment by biswassri** [comment]
@MarianMacik The PR check finally passed. Let me know if think any other changes are required.

**Review by MarianMacik — APPROVED** [review]
/approve

**Pull request merged**
